### PR TITLE
[SPARK-10215]  [SQL] Fix precision of division (follow the rule in Hive)

### PR DESCRIPTION
**Pull request by davies**
Follow the rule in Hive for decimal division. see https://github.com/apache/hive/blob/ac755ebe26361a4647d53db2a28500f71697b276/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFOPDivide.java#L113

cc @chenghao-intel

### CODE REVIEW

**Comment by chenghao-intel** [comment]
I've tested with some corner case, it works for me. LGTM, except the unit test itself need to update with the new `scale`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 1]
```diff
@@ -400,8 +400,15 @@ object HiveTypeCoercion {
...
         case Divide(e1 @ DecimalType.Expression(p1, s1), e2 @ DecimalType.Expression(p2, s2)) =>
-          val resultType = DecimalType.bounded(p1 - s1 + s2 + max(6, s1 + p2 + 1),
-            max(6, s1 + p2 + 1))
+          var intDig = min(DecimalType.MAX_SCALE, p1 - s1 + s2)
+          var decDig = min(DecimalType.MAX_SCALE, max(6, s1 + p2 + 1))
+          val diff = (intDig + decDig) - DecimalType.MAX_SCALE
+          if (diff > 0) {
+            decDig -= diff / 2 + 1
+            intDig = DecimalType.MAX_SCALE - decDig
+            DecimalType.bounded(intDig + decDig, decDig)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: This line is not used. I will remove it when I merge the pr.

**Comment by yhuai** [comment]
LGTM. Merging to master and branch 1.5.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 2]
```diff
@@ -400,8 +400,15 @@ object HiveTypeCoercion {
...
+          }
+          val resultType = DecimalType.bounded(intDig + decDig, decDig)
           val widerType = widerDecimalType(p1, s1, p2, s2)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: is the `resultType` always tighter than `widerType`?

**Pull request closed without merging**